azurelinuxagent/pa/provision/cloudinit.py

```diff
@@ -132,3 +126,43 @@ def wait_for_ssh_host_key(self, max_retry=1800, sleep_time=1):
         raise ProvisionError("Giving up, ssh host key was not found at {0} "
                              "after {1}s".format(path,
                                                  max_retry * sleep_time))
+
+def cloud_init_is_enabled():
+    """
+    Determine whether or not cloud-init is enabled.
+
+    Args:
+        None
+
+    Returns:
+        bool - True if cloud-init is enabled, False if otherwise.
+    """
+
+    unit_name = 'cloud-init-local.service'
+
+    bus = dbus.SystemBus()
+    systemd_dbus = bus.get_object(
+        'org.freedesktop.systemd1',
+        '/org/freedesktop/systemd1'
+    )
+    systemd_manager = dbus.Interface(
+        systemd_dbus,
+        'org.freedesktop.systemd1.Manager'
+    )
+
+    try:
+        unit_is_enabled = systemd_manager.GetUnitFileState(unit_name).lower() == 'enabled'
+    except dbus.DBusException as exc:
+        # If the unit does not exist, then take that as a falsy
+        # condition. We don't want to throw an exception in this
+        # case because waagent should be able to handle the situation
+        # where there is no cloud-init installed.
+        if exc.get_dbus_name() == 'org.freedesktop.DBus.Error.FileNotFound':
+            unit_is_enabled = False
+        else:
+            # Any other dbus exception should be bubbled up though.
+            logger.error('Unexpected dbus error: {}'.format(exc))
```

Comment:
Why do we need to raise the error here and fail the provisioning. Can't we just consider this as if the cloud-init is not enabled and fall back to waagent ?

Reply:
That's exactly one of the changes I'm making now. I don't think any code path in this should fail provisioning at all. I think any exception is just a falsy return with no exception. Agreed.